pkg/component/component.go

```diff
@@ -125,6 +125,143 @@ func ApplyConfig(client kclient.ClientInterface, envSpecificInfo envinfo.EnvSpec
+func RemoveDuplicateComponentsForListingOutput(components []Component) []Component {
+
+	var outputComp []Component
+
+	// Loop through all the components
+	for _, component := range components {
+		found := false
+		for _, check := range outputComp {
+
+			// Make the variable names a bit shorter for readability
+			componentLabel, checkLabel := component.ObjectMeta.Labels, check.ObjectMeta.Labels
+
+			// If we find ones that match the same instance label
+			if componentLabel[componentlabels.ComponentKubernetesInstanceLabel] == checkLabel[componentlabels.ComponentKubernetesInstanceLabel] {
+
+				// Mark as found
+				found = true
+
+				// Sometimes Kubernetes operators deploy an ingress / service / etc via an operator but do NOT include a
+				// "app.kubernetes.io/managed-by" label. So when we "deduplicate" the component list, this may appear blank.
+				// If it's blank, and there is an instance of the same name with a managed-by label, we propagate that to the output
+				// components
+				if componentLabel[componentlabels.ComponentManagedByLabel] != "" && checkLabel[componentlabels.ComponentManagedByLabel] == "" {
+					checkLabel[componentlabels.ComponentManagedByLabel] = componentLabel[componentlabels.ComponentManagedByLabel]
+				}
+
+				// Check if there are TWO components, one in dev, one in deploy. If so, we will keep the *first* component, and
+				// update the mode label to indicate it's running in both dev and deploy.
+				// We check this *explicitely* in order to keep the correct order when displaying (ex. Dev, Deploy)
+				// instead of some components showing Deploy, Dev depending on the order that the components are in the array.
+				// We also use this long if statement in case there is something odd that went on (user modified the label outside of odo)
+				componentMode := componentLabel[componentlabels.ComponentModeLabel]
+				checkMode := checkLabel[componentlabels.ComponentModeLabel]
+				if (componentMode != checkMode) &&
+					(componentMode == componentlabels.ComponentDevLabel || componentMode == componentlabels.ComponentDeployLabel) &&
+					(checkMode == componentlabels.ComponentDevLabel || checkMode == componentlabels.ComponentDeployLabel) {
+					checkLabel[componentlabels.ComponentModeLabel] = fmt.Sprintf("%s, %s", componentlabels.ComponentDevLabel, componentlabels.ComponentDeployLabel)
+				}
+
+				// Set the labels in the end
+				check.SetLabels(checkLabel)
+			}
+		}
+
+		// If not found, append to the component output
+		if !found {
+			outputComp = append(outputComp, component)
+		}
+
+	}
+
+	return outputComp
+}
+
+// ListAllClusterComponents returns a list of all "components" on a cluster
+// that are both odo and non-odo components.
+//
+// We then return a list of "components" intended for listing / output purposes specifically for commands such as:
+// `odo list`
+func ListAllClusterComponents(client kclient.ClientInterface, namespace string) ([]Component, error) {
```

Comment:
Because this method is more the CLI area / rather than business logic, I would IMO keep it here. Only reasoning is because we curate the output of []Component that's specific for listing purposes. So it wouldn't be ideal to put it into the interface / client method since this function is more of a "one time use" purpose.

Reply:
Almost all the methods that you see in these interfaces are specific to the CLI layer, thereby a mainly "one time use" purpose IMO. Although, that does make me question if we are going about this Client business in the right way.
@feloy care to chime in?